WebGL 2.0 Support

gchoqueux opened the pull request:
## Description

- [x] **WebGl 2.0 support**
   - [x] `webGLRenderer#render` throws error:
      - [x] **_WebGL: INVALID_OPERATION: drawElements: no buffer is bound to enabled attribute_**
        Fixed with removing common attribute in before dispose geometry;
      - [x] With Chrome v85 **THREE.WebGLProgram: gl.getProgramInfoLog() return string with null char**
      - [x] Another undefined webGL error. This could be the same cause.          
              on Firefox/ubuntu : **WebGL warning: drawElementsInstanced: Index buffer not bound.** 
              on Safari/macOs : **INVALID_OPERATION: drawElements: no ELEMENT_ARRAY_BUFFER bound**
   - [x] `RawShaderMaterial` and _GLSL 3.0_ conversion;
     - problem with `RawShaderMaterial` to add version and special defines before auto defines mechanism;
   - [x] avoid undefined uniforms;
   - [x] fix unit tests fail;
   - [x] fix point cloud shader;
   - [x] fix `DataTexture` instance with float array: use `RedFormat` and `R32F` for internal format;
   - [x] avoid `#pragma unrool` with WebGL 2.0; => **it's same problem with WebGl 2.0**
   - [x]  fix `OrientedImageMaterial`;

fix #1444

gchoqueux committed:
refactor(picking): avoid to instance typed array for each full depth buffer.

gchoqueux committed:
refactor(geometry): adjust index buffer size to vertex count.

gchoqueux committed:
fix(test): add default value to tile segment.

gchoqueux committed:
chore: update THREE package.

zarov commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -54,6 +67,12 @@ function updateLayersUniforms(uniforms, olayers, max) {
     }
     textureCount.value = count;
+
+    for (let i = count; i < textures.length; i++) {
```
What is the purpose of this ?

gchoqueux replied:
WebGL 2.0 doesn't support the undefined uniforms, so they are defined by default value.

zarov commented on src/Core/Prefab/computeBufferTileGeometry.js:
```diff
@@ -43,7 +47,13 @@ export default function computeBuffers(params) {
     computeUvs[0] = () => {};
     if (params.buildIndexAndUv_0) {
-        outBuffers.index = new Uint32Array(triangles * 3);
+        if (nVertex < 2 ** 8) {
+            outBuffers.index = new Uint8Array(triangles * 3);
+        } else if (nVertex < 2 ** 16) {
+            outBuffers.index = new Uint16Array(triangles * 3);
+        } else if (nVertex < 2 ** 32) {
+            outBuffers.index = new Uint32Array(triangles * 3);
+        }
```
Well seen !

zarov commented on src/Renderer/c3DEngine.js:
```diff
@@ -62,7 +66,7 @@ class c3DEngine {
             viewerDiv.appendChild(this.label2dRenderer.domElement);
 
             // WIP: support WebGL 2.0 see https://github.com/iTowns/itowns/pull/1443
```
This comment can be discarded

gchoqueux replied:
removed

zarov commented on test/unit/bootstrap.js:
```diff
@@ -15,6 +15,8 @@ global.requestAnimationFrame = () => {};
 global.fetch = fetch;
 global.fetch.Promise = Promise;
 
+global.navigator = undefined;
```
Side note: this could be replaced by `jsdom.Navigator` in #1412

gchoqueux replied:
I add a comment

zarov commented on src/Core/Prefab/TileBuilder.js:
```diff
@@ -25,38 +25,47 @@ export default function newTileGeometry(builder, params) {
+                // to avoid remove index buffer and attribute buffer uv_0
+                // error un-bound buffer in webgl with VAO rendering
+                geometry.index = null;
+                delete geometry.attributes.uv_0;
```
Could you leave a note for posterity, to change this, if there is a fix some times in THREE ?

gchoqueux replied:
comment added

zarov commented on src/Core/Prefab/TileBuilder.js:
```diff
@@ -25,38 +25,47 @@ export default function newTileGeometry(builder, params) {
+                cacheTile.delete(south, params.level, bufferKey);
+            }
+        };
+        resolve(geometry);
+        return Promise.resolve({ geometry, quaternion, position });
```
Doesn't this cause any sync problem ? I don't understand the "double" resolving.

gchoqueux replied:
the first resolving resolves the cached promise.
the second resolving resolves the returned `newTileGeometry` promise.

zarov commented on src/Core/View.js:
```diff
@@ -188,14 +188,8 @@ class View extends THREE.EventDispatcher {
-                this._fullSizeDepthBuffer = null;
-            }
-        });
+        this._fullSizeDepthBuffer = new Uint8Array(4 * this.camera.width * this.camera.height);
```
In `notifyChange` you can remove the check on this variable.

gchoqueux replied:
yes, good catch!

gchoqueux replied:
removed

zarov commented on src/Renderer/Shader/Chunk/WebGL2_pars_fragment.glsl:
```diff
@@ -0,0 +1,25 @@
+#if defined(WEBGL2)
```
Could you link the `WebGLProgram` source on this ?

gchoqueux replied:
comment added

zarov commented on src/Renderer/Shader/Chunk/WebGL2_pars_fragment.glsl:
```diff
@@ -0,0 +1,25 @@
+#define textureCubeLodEXT textureLod
+#define texture2DGradEXT textureGrad
+#define texture2DProjGradEXT textureProjGrad
+#define textureCubeGradEXT textureGrad
```
This is in double

gchoqueux replied:
yes!

gchoqueux replied:
double removed

zarov commented on src/Renderer/Shader/Chunk/elevation_pars_vertex.glsl:
```diff
@@ -5,6 +5,7 @@
         int mode;
         float zmin;
         float zmax;
+        int crs;
```
You forgot to remove this.

gchoqueux replied:
yes

gchoqueux replied:
removed

zarov commented on src/Renderer/Shader/PointsFS.glsl:
```diff
@@ -1,9 +1,15 @@
 #endif
 
+#ifdef GL_OES_standard_derivatives
+#extension GL_OES_standard_derivatives : enable
```
We don't seem to use this extension, is this needed ?

gchoqueux replied:
i remove it

zarov reviewed (commented):
Thanks for this, a few comments but it should be good to go today ;)

gchoqueux committed:
feature(Webgl): add support WebGl 2.0.

* WebGL: INVALID_OPERATION: drawElements: no buffer is bound to enabled attribute
   => fix : avoid shared buffer attribute deleting
* RawShaderMaterial/GLSL 3.0 conversion problem with RawShaderMaterial
  to add version and special defines before auto defines mechanism.
   => fixed with Material.onBeforeCompile
* set default uniforms value.

gchoqueux committed:
refactor(test): add unit tests and error handling.
